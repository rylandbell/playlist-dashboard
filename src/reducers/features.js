const initialFeatureValues = [
  {
    name: 'danceability',
    isActive: true,
    isGraphed: true,
    isDim: false,
    sortBy: false,
    showReferenceLine: false,
    currentValue: [0, 1]
  },
  {
    name: 'energy',
    isActive: true,
    isGraphed: false,
    isDim: false,
    sortBy: false,
    showReferenceLine: false,
    currentValue: [0, 1]
  },
  {
    name: 'valence',
    isActive: true,
    isGraphed: false,
    isDim: false,
    sortBy: false,
    showReferenceLine: false,
    currentValue: [0, 1]
  },
  {
    name: 'popularity',
    isActive: true,
    isGraphed: false,
    isDim: false,
    sortBy: false,
    showReferenceLine: false,
    currentValue: [0, 1]
  },
  {
    name: 'acousticness',
    isActive: true,
    isGraphed: false,
    isDim: false,
    sortBy: false,
    showReferenceLine: false,
    currentValue: [0, 1]
  },
  {
    name: 'instrumentalness',
    isActive: true,
    isGraphed: false,
    isDim: false,
    sortBy: false,
    showReferenceLine: false,
    currentValue: [0, 1]
  }
  // {
  //   name: 'liveness',
  //   isActive: true,
  //   isGraphed: false,
  // isDim: false,
  //   showReferenceLine: false,
  //   currentValue: [0,1],
  // },
];

export const features = (state = initialFeatureValues, action) => {
  let targetFeature, updatedFeature;
  switch (action.type) {
    case 'UPDATE_FEATURE_FILTER':
      targetFeature = state[action.featureIndex];
      updatedFeature = { ...targetFeature, currentValue: action.data };
      return [
        ...state.slice(0, action.featureIndex),
        updatedFeature,
        ...state.slice(action.featureIndex + 1)
      ];
    case 'TOGGLE_CHARTED_FEATURE':
      return state.map((feature, index) => {
        if (index === action.featureIndex) {
          return { ...feature, isGraphed: action.newValue };
        } else {
          return feature;
        }
      });
    case 'START_DRAGGING_FEATURE_SLIDER':
      return state.map((feature, index) => {
        if (index === action.featureIndex) {
          return { ...feature, showReferenceLine: true, isGraphed: true };
        } else {
          return { ...feature, isDim: true };
        }
      });
    case 'STOP_DRAGGING_FEATURE_SLIDER':
      return state.map(feature => ({
        ...feature,
        showReferenceLine: false,
        isDim: false
      }));
    case 'SORT_BY_FEATURE':
      return state.map(feature => {
        if (feature.name === action.payload.featureName) {
          
          // cycle through 'ascending', 'descending', false:
          let newSortByValue
          if (feature.sortBy === 'ascending') {
            newSortByValue = 'descending';
          } else if (feature.sortBy === 'descending') {
            newSortByValue = false;
          } else {
            newSortByValue = 'ascending';
          }

          return {...feature, sortBy: newSortByValue};

        //set all other features to unsorted:
        } else {
          return {...feature, sortBy: false};
        }
      });
    default:
      return state;
  }
};
