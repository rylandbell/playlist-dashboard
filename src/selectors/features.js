//combine unchanging data about features with their current state

import { createSelector } from 'reselect';
import { staticFeaturesData } from '../constants/staticFeaturesData';

const getFeaturesFromStore = state => state.features;

export const getFeaturesData = createSelector([getFeaturesFromStore], stateFeatures => {
  return stateFeatures.map(stateFeature => {
    // find the corresponding feature in staticFeaturesData (look for name match):
    const match = staticFeaturesData.find(
      feature => feature.name === stateFeature.name
    );
    return { ...stateFeature, ...match };
  });
});
