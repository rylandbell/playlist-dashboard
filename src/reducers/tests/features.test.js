import deepFreeze from "deep-freeze";
import * as reducers from "../features";

describe("Features reducers", () => {
  const initState = reducers.features(undefined, {
    type: "INIT"
  });

  it("should update filter value for a single feature with UPDATE_FEATURE_FILTER actions", () => {
    const action = {
      type: "UPDATE_FEATURE_FILTER",
      featureIndex: 0,
      data: [0.1, 0.2]
    };

    deepFreeze(initState);
    deepFreeze(action);

    expect(reducers.features(initState, action)).toMatchSnapshot();
  });

  it("should update isCharted value for a single feature with TOGGLE_CHARTED_FEATURE actions", () => {
    const action = {
      type: "TOGGLE_CHARTED_FEATURE",
      featureIndex: 2,
      newValue: true
    };

    deepFreeze(initState);
    deepFreeze(action);

    expect(reducers.features(initState, action)).toMatchSnapshot();
  });

  it("should showReferenceLine for a single feature, and set all other featuers to isDim, with START_DRAGGING_FEATURE_SLIDER actions", () => {
    const action = {
      type: "START_DRAGGING_FEATURE_SLIDER",
      featureIndex: 4
    };

    deepFreeze(initState);
    deepFreeze(action);

    expect(reducers.features(initState, action)).toMatchSnapshot();
  });

  it("should set all isDim and showReferenceLine values to false for a single feature with STOP_DRAGGING_FEATURE_SLIDER actions", () => {
    //first send a START_DRAGGING action to update the state
    const action1 = {
      type: "START_DRAGGING_FEATURE_SLIDER",
      featureIndex: 4
    };
    const draggingState = reducers.features(initState, action1);

    const action2 = {
      type: "STOP_DRAGGING_FEATURE_SLIDER"
    };

    deepFreeze(draggingState);
    deepFreeze(action2);

    expect(reducers.features(draggingState, action2)).toMatchSnapshot();
  });

  it('feature\'s sortBy value should go from false to \"ascending\" for SORT_BY_FILTER actions', () => {
    const action = {
      type: "SORT_BY_FEATURE",
      payload: {
        featureName: 'energy'
      }
    };

    deepFreeze(initState);
    deepFreeze(action);

    expect(reducers.features(initState, action)).toMatchSnapshot();
  });

  it('feature\'s sortBy value should go from \"ascending\" to \"descending\" for SORT_BY_FILTER actions', () => {
    const action = {
      type: "SORT_BY_FEATURE",
      payload: {
        featureName: 'energy'
      }
    };

    //apply the action once before testing to set sortBy to "ascending";
    const ascendingState = reducers.features(initState, action);
    deepFreeze(ascendingState);
    deepFreeze(action);

    expect(reducers.features(ascendingState, action)).toMatchSnapshot();
  });

  it('feature\'s sortBy value should go from \"descending\" to false for SORT_BY_FILTER actions', () => {
    const action = {
      type: "SORT_BY_FEATURE",
      payload: {
        featureName: 'energy'
      }
    };

    //apply the action twice before testing to set sortBy to "ascending";
    const ascendingState = reducers.features(initState, action);
    const descendingState = reducers.features(ascendingState, action);
    deepFreeze(descendingState);
    deepFreeze(action);

    expect(reducers.features(descendingState, action)).toMatchSnapshot();
  });

  it('features not included in payload should have sortBy values set to false for SORT_BY_FILTER actions', () => {
    const action1 = {
      type: "SORT_BY_FEATURE",
      payload: {
        featureName: 'energy'
      }
    };
    const sortedState = reducers.features(initState, action1);

    const action2 = {
      type: "SORT_BY_FEATURE",
      payload: {
        featureName: 'popularity'
      }
    };

    deepFreeze(sortedState);
    deepFreeze(action2);

    expect(reducers.features(sortedState, action2)).toMatchSnapshot();
  });
});
