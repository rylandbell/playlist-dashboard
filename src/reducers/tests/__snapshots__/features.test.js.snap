// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Features reducers feature's sortBy value should go from "ascending" to "descending" for SORT_BY_FILTER actions 1`] = `
Array [
  Object {
    "currentValue": Array [
      0,
      1,
    ],
    "isActive": true,
    "isDim": false,
    "isGraphed": true,
    "name": "danceability",
    "showReferenceLine": false,
    "sortBy": false,
  },
  Object {
    "currentValue": Array [
      0,
      1,
    ],
    "isActive": true,
    "isDim": false,
    "isGraphed": false,
    "name": "energy",
    "showReferenceLine": false,
    "sortBy": "descending",
  },
  Object {
    "currentValue": Array [
      0,
      1,
    ],
    "isActive": true,
    "isDim": false,
    "isGraphed": false,
    "name": "valence",
    "showReferenceLine": false,
    "sortBy": false,
  },
  Object {
    "currentValue": Array [
      0,
      1,
    ],
    "isActive": true,
    "isDim": false,
    "isGraphed": false,
    "name": "popularity",
    "showReferenceLine": false,
    "sortBy": false,
  },
  Object {
    "currentValue": Array [
      0,
      1,
    ],
    "isActive": true,
    "isDim": false,
    "isGraphed": false,
    "name": "acousticness",
    "showReferenceLine": false,
    "sortBy": false,
  },
  Object {
    "currentValue": Array [
      0,
      1,
    ],
    "isActive": true,
    "isDim": false,
    "isGraphed": false,
    "name": "instrumentalness",
    "showReferenceLine": false,
    "sortBy": false,
  },
]
`;

exports[`Features reducers feature's sortBy value should go from "descending" to false for SORT_BY_FILTER actions 1`] = `
Array [
  Object {
    "currentValue": Array [
      0,
      1,
    ],
    "isActive": true,
    "isDim": false,
    "isGraphed": true,
    "name": "danceability",
    "showReferenceLine": false,
    "sortBy": false,
  },
  Object {
    "currentValue": Array [
      0,
      1,
    ],
    "isActive": true,
    "isDim": false,
    "isGraphed": false,
    "name": "energy",
    "showReferenceLine": false,
    "sortBy": false,
  },
  Object {
    "currentValue": Array [
      0,
      1,
    ],
    "isActive": true,
    "isDim": false,
    "isGraphed": false,
    "name": "valence",
    "showReferenceLine": false,
    "sortBy": false,
  },
  Object {
    "currentValue": Array [
      0,
      1,
    ],
    "isActive": true,
    "isDim": false,
    "isGraphed": false,
    "name": "popularity",
    "showReferenceLine": false,
    "sortBy": false,
  },
  Object {
    "currentValue": Array [
      0,
      1,
    ],
    "isActive": true,
    "isDim": false,
    "isGraphed": false,
    "name": "acousticness",
    "showReferenceLine": false,
    "sortBy": false,
  },
  Object {
    "currentValue": Array [
      0,
      1,
    ],
    "isActive": true,
    "isDim": false,
    "isGraphed": false,
    "name": "instrumentalness",
    "showReferenceLine": false,
    "sortBy": false,
  },
]
`;

exports[`Features reducers feature's sortBy value should go from false to "ascending" for SORT_BY_FILTER actions 1`] = `
Array [
  Object {
    "currentValue": Array [
      0,
      1,
    ],
    "isActive": true,
    "isDim": false,
    "isGraphed": true,
    "name": "danceability",
    "showReferenceLine": false,
    "sortBy": false,
  },
  Object {
    "currentValue": Array [
      0,
      1,
    ],
    "isActive": true,
    "isDim": false,
    "isGraphed": false,
    "name": "energy",
    "showReferenceLine": false,
    "sortBy": "ascending",
  },
  Object {
    "currentValue": Array [
      0,
      1,
    ],
    "isActive": true,
    "isDim": false,
    "isGraphed": false,
    "name": "valence",
    "showReferenceLine": false,
    "sortBy": false,
  },
  Object {
    "currentValue": Array [
      0,
      1,
    ],
    "isActive": true,
    "isDim": false,
    "isGraphed": false,
    "name": "popularity",
    "showReferenceLine": false,
    "sortBy": false,
  },
  Object {
    "currentValue": Array [
      0,
      1,
    ],
    "isActive": true,
    "isDim": false,
    "isGraphed": false,
    "name": "acousticness",
    "showReferenceLine": false,
    "sortBy": false,
  },
  Object {
    "currentValue": Array [
      0,
      1,
    ],
    "isActive": true,
    "isDim": false,
    "isGraphed": false,
    "name": "instrumentalness",
    "showReferenceLine": false,
    "sortBy": false,
  },
]
`;

exports[`Features reducers features not included in payload should have sortBy values set to false for SORT_BY_FILTER actions 1`] = `
Array [
  Object {
    "currentValue": Array [
      0,
      1,
    ],
    "isActive": true,
    "isDim": false,
    "isGraphed": true,
    "name": "danceability",
    "showReferenceLine": false,
    "sortBy": false,
  },
  Object {
    "currentValue": Array [
      0,
      1,
    ],
    "isActive": true,
    "isDim": false,
    "isGraphed": false,
    "name": "energy",
    "showReferenceLine": false,
    "sortBy": false,
  },
  Object {
    "currentValue": Array [
      0,
      1,
    ],
    "isActive": true,
    "isDim": false,
    "isGraphed": false,
    "name": "valence",
    "showReferenceLine": false,
    "sortBy": false,
  },
  Object {
    "currentValue": Array [
      0,
      1,
    ],
    "isActive": true,
    "isDim": false,
    "isGraphed": false,
    "name": "popularity",
    "showReferenceLine": false,
    "sortBy": "ascending",
  },
  Object {
    "currentValue": Array [
      0,
      1,
    ],
    "isActive": true,
    "isDim": false,
    "isGraphed": false,
    "name": "acousticness",
    "showReferenceLine": false,
    "sortBy": false,
  },
  Object {
    "currentValue": Array [
      0,
      1,
    ],
    "isActive": true,
    "isDim": false,
    "isGraphed": false,
    "name": "instrumentalness",
    "showReferenceLine": false,
    "sortBy": false,
  },
]
`;

exports[`Features reducers should set all isDim and showReferenceLine values to false for a single feature with STOP_DRAGGING_FEATURE_SLIDER actions 1`] = `
Array [
  Object {
    "currentValue": Array [
      0,
      1,
    ],
    "isActive": true,
    "isDim": false,
    "isGraphed": true,
    "name": "danceability",
    "showReferenceLine": false,
    "sortBy": false,
  },
  Object {
    "currentValue": Array [
      0,
      1,
    ],
    "isActive": true,
    "isDim": false,
    "isGraphed": false,
    "name": "energy",
    "showReferenceLine": false,
    "sortBy": false,
  },
  Object {
    "currentValue": Array [
      0,
      1,
    ],
    "isActive": true,
    "isDim": false,
    "isGraphed": false,
    "name": "valence",
    "showReferenceLine": false,
    "sortBy": false,
  },
  Object {
    "currentValue": Array [
      0,
      1,
    ],
    "isActive": true,
    "isDim": false,
    "isGraphed": false,
    "name": "popularity",
    "showReferenceLine": false,
    "sortBy": false,
  },
  Object {
    "currentValue": Array [
      0,
      1,
    ],
    "isActive": true,
    "isDim": false,
    "isGraphed": true,
    "name": "acousticness",
    "showReferenceLine": false,
    "sortBy": false,
  },
  Object {
    "currentValue": Array [
      0,
      1,
    ],
    "isActive": true,
    "isDim": false,
    "isGraphed": false,
    "name": "instrumentalness",
    "showReferenceLine": false,
    "sortBy": false,
  },
]
`;

exports[`Features reducers should showReferenceLine for a single feature, and set all other featuers to isDim, with START_DRAGGING_FEATURE_SLIDER actions 1`] = `
Array [
  Object {
    "currentValue": Array [
      0,
      1,
    ],
    "isActive": true,
    "isDim": true,
    "isGraphed": true,
    "name": "danceability",
    "showReferenceLine": false,
    "sortBy": false,
  },
  Object {
    "currentValue": Array [
      0,
      1,
    ],
    "isActive": true,
    "isDim": true,
    "isGraphed": false,
    "name": "energy",
    "showReferenceLine": false,
    "sortBy": false,
  },
  Object {
    "currentValue": Array [
      0,
      1,
    ],
    "isActive": true,
    "isDim": true,
    "isGraphed": false,
    "name": "valence",
    "showReferenceLine": false,
    "sortBy": false,
  },
  Object {
    "currentValue": Array [
      0,
      1,
    ],
    "isActive": true,
    "isDim": true,
    "isGraphed": false,
    "name": "popularity",
    "showReferenceLine": false,
    "sortBy": false,
  },
  Object {
    "currentValue": Array [
      0,
      1,
    ],
    "isActive": true,
    "isDim": false,
    "isGraphed": true,
    "name": "acousticness",
    "showReferenceLine": true,
    "sortBy": false,
  },
  Object {
    "currentValue": Array [
      0,
      1,
    ],
    "isActive": true,
    "isDim": true,
    "isGraphed": false,
    "name": "instrumentalness",
    "showReferenceLine": false,
    "sortBy": false,
  },
]
`;

exports[`Features reducers should update filter value for a single feature with UPDATE_FEATURE_FILTER actions 1`] = `
Array [
  Object {
    "currentValue": Array [
      0.1,
      0.2,
    ],
    "isActive": true,
    "isDim": false,
    "isGraphed": true,
    "name": "danceability",
    "showReferenceLine": false,
    "sortBy": false,
  },
  Object {
    "currentValue": Array [
      0,
      1,
    ],
    "isActive": true,
    "isDim": false,
    "isGraphed": false,
    "name": "energy",
    "showReferenceLine": false,
    "sortBy": false,
  },
  Object {
    "currentValue": Array [
      0,
      1,
    ],
    "isActive": true,
    "isDim": false,
    "isGraphed": false,
    "name": "valence",
    "showReferenceLine": false,
    "sortBy": false,
  },
  Object {
    "currentValue": Array [
      0,
      1,
    ],
    "isActive": true,
    "isDim": false,
    "isGraphed": false,
    "name": "popularity",
    "showReferenceLine": false,
    "sortBy": false,
  },
  Object {
    "currentValue": Array [
      0,
      1,
    ],
    "isActive": true,
    "isDim": false,
    "isGraphed": false,
    "name": "acousticness",
    "showReferenceLine": false,
    "sortBy": false,
  },
  Object {
    "currentValue": Array [
      0,
      1,
    ],
    "isActive": true,
    "isDim": false,
    "isGraphed": false,
    "name": "instrumentalness",
    "showReferenceLine": false,
    "sortBy": false,
  },
]
`;

exports[`Features reducers should update isCharted value for a single feature with TOGGLE_CHARTED_FEATURE actions 1`] = `
Array [
  Object {
    "currentValue": Array [
      0,
      1,
    ],
    "isActive": true,
    "isDim": false,
    "isGraphed": true,
    "name": "danceability",
    "showReferenceLine": false,
    "sortBy": false,
  },
  Object {
    "currentValue": Array [
      0,
      1,
    ],
    "isActive": true,
    "isDim": false,
    "isGraphed": false,
    "name": "energy",
    "showReferenceLine": false,
    "sortBy": false,
  },
  Object {
    "currentValue": Array [
      0,
      1,
    ],
    "isActive": true,
    "isDim": false,
    "isGraphed": true,
    "name": "valence",
    "showReferenceLine": false,
    "sortBy": false,
  },
  Object {
    "currentValue": Array [
      0,
      1,
    ],
    "isActive": true,
    "isDim": false,
    "isGraphed": false,
    "name": "popularity",
    "showReferenceLine": false,
    "sortBy": false,
  },
  Object {
    "currentValue": Array [
      0,
      1,
    ],
    "isActive": true,
    "isDim": false,
    "isGraphed": false,
    "name": "acousticness",
    "showReferenceLine": false,
    "sortBy": false,
  },
  Object {
    "currentValue": Array [
      0,
      1,
    ],
    "isActive": true,
    "isDim": false,
    "isGraphed": false,
    "name": "instrumentalness",
    "showReferenceLine": false,
    "sortBy": false,
  },
]
`;
